### PR TITLE
Addition of all zoofs  algos 

### PR DESCRIPTION
**Pull request by jaswinder9051998**
## Related Issue or bug

Info about Issue or bug.

Fixes: #[issue number that will be closed through this PR]

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Code style update (formatting, local variables).
- [ ] Breaking change (fix or feature that would cause existing functionality to stop working).
- [ ] This change requires a documentation update.

## Describe the changes you've made

added all zoofs algos to FeatureSelector's fit

## How has this been tested?

Changes for the new code was covered in the following
tests\test_feature_engineering
- test_solver_parameter_empty_zoofs
- test_transform_zoofs
- test_objective_function_zoofs

## Is there any unusual behaviour in your code?

None noted

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

## Screenshots

Original | Updated
:--------------------: | :--------------------:
**original screenshot** | **updated screenshot** |

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (tvdboom) [inline .vscode/settings.json:1]: Please remove this file from git tracking. It should not be added to the repo.

File: atom/feature_engineering.py [hunk 2]
```diff
@@ -988,12 +1014,68 @@ def check_y():
...
+            if ("X_valid" in self.kwargs.keys()) and ("y_valid" in self.kwargs.keys()): # catching X_valid if provided
+                X_valid, y_valid = self._prepare_input(
+                    self.kwargs["X_valid"], self.kwargs["y_valid"])
+            else:
+                X_valid, y_valid = X, y
+
+            self.algo.fit(model=self._solver,
+                          X_train=X,
+                          y_train=y,
+                          X_valid=X_valid,
```
Comment (tvdboom) [inline atom/feature_engineering.py:1068]: If the user doesn't provide a validation set, it validates on the training set (since X_valid = X)? I am not sure if this is the right approach since the results of the objective_function will be 100% biased and the algorithm will be prone to overfitting. The way to go is probably making a split or cross-validation over X and feed those in X_valid. Similar to what sklearn does in RFE.

File: atom/feature_engineering.py [hunk 3]
```diff
@@ -988,12 +1014,68 @@ def check_y():
...
+                          y_valid=y_valid,
+                          verbose=False if self.verbose == 0 else True)
```
Comment (tvdboom) [inline atom/feature_engineering.py:1070]: Should be `if self.verbose > 1 else`. ATOM prints the output of external libraries only for verbose=2.

File: atom/feature_engineering.py [hunk 4]
```diff
@@ -1146,4 +1231,15 @@ def transform(self, X: X_TYPES, y: Optional[Y_TYPES] = None):
             idx = np.argsort(get_feature_importance(self.rfecv.estimator_))
             self.feature_importance = list(X.columns[idx][::-1])
 
+        elif self.strategy.lower() in ["pso", "gwo", "dfo", "geneo", "hho"]:
```
Comment (tvdboom) [inline atom/feature_engineering.py:1234]: It would be nice if we can add the `self.feature_importance` attribute here as well. Are the `algo.best_feature_list` features ordered by importance or in any defined order?

**Comment by tvdboom** [comment]
Thanks for the PR. I added some comments you can have a look at. To resolve the conflict, merge the development branch from origin into your branch.

File: atom/feature_engineering.py (hunk 2 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py:1068]: For zoofs optimization,  validation set and objective_function are prerequisites so that the user has the flexibility to perform their own operation under 'objective function'. The reasoning for not forcing a cv split over the xtrain (if validation set isn't provided) was that some datasets which are time-dependant (eg basketball matches, tournaments  ) might require a TimeSeriesSplit. In such senarios the user in their own wisdom can choose what to perform under   'objective function' themselves. 

For atom, the two cases where 'objective_function' and 'X_valid' aren't explicitly provided, are the only situations where we can enforce a cv split. i,e 
- Case 1 : where user only provides 'scoring'
- Case 2 : where user provides nothing

if you agree with the above discussion, i can proceed

File: atom/feature_engineering.py (hunk 3 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py:1070]: thanks for pointing it out, will make the change

File: atom/feature_engineering.py (hunk 4 above)
Comment (jaswinder9051998) [reply atom/feature_engineering.py:1234]: All algos are wrapper algo which inherently work on exploration and exploitation of the search space. There isn't any component of feature_importance involved in the algos

Comment (tvdboom) [reply atom/feature_engineering.py:1234]: Oke, then we leave it as it is now

File: atom/feature_engineering.py [hunk 5]
```diff
@@ -988,12 +1014,68 @@ def check_y():
...
+            if self.kwargs.get("objective_function"): # zoofs-native objective_function choice
+                objective_function = self.kwargs['objective_function']
+
+            elif self.kwargs.get("scoring"): # atom-native scoring method can be chosen 
+                params_for_objective_function = {
+                    "scorer": get_custom_scorer(self.kwargs["scoring"])}
+                objective_function = custom_function_for_scorer
+
+            else: # default scoring method is chosen if nothing is provided
+                params_for_objective_function = {"scorer": get_custom_scorer("neg_log_loss") if hasattr(
```
Comment (tvdboom) [inline atom/feature_engineering.py:1047]: atom uses default scorers `f1` for binary classification, `f1_weighted` for multiclass classification and `r2` for regression. You can check which one it is with the attribute `goal`. I think it's best to change them here to atom's default values for consistency

File: atom/feature_engineering.py (hunk 2 above)
Comment (tvdboom) [reply atom/feature_engineering.py:1068]: Yes, that sounds good. I expect the majority of users not to add any of those two parameters
